### ISSUE TITLE
cmd/go: change tag can not get the new commit version

### ISSUE THREAD

**Issue by iiinsomnia** [golang/go#31903]
now, there is a go module A.

step1: add new tag `v1.0.1` to `commit_1`
step2: delete the tag `v1.0.1`
step3: modify and commit as `commit_2`
step4: add the tag `v1.0.1` to `commit_2`

when I run `go get A@v1.0.1`, it gets the version of `commit_1` instead of `commit_2` and in the `go.mod` shows `A v0.0.0-20190507092345-afa2bget21xc` instead of `A v1.0.1`

**Comment by andybons** [comment]
@iiinsomnia what version of Go are you using? What's your environment? In other words, please use the new issue template.

@bcmills @jayconrod

**Comment by thepudds** [comment]
@iiinsomnia FWIW, Go modules require that modules follow the [semver](https://semver.org) spec, and as far as I am aware, it is not allowed under the semver spec to move a tag from commit_1 to commit_2 as in your example if that initial tagging represents an actual release viewable by others.

For example, from https://semver.org/#spec-item-3:

> 3. Once a versioned package has been released, the contents of that version MUST NOT be modified. Any modifications MUST be released as a new version.

That said, in your case, you might want to try doing `go clean -modcache`. That might help, but not sure.

Also, the top-half of your example is using v1 major version tags, but then the last part of your example says:
> go get A@v0.0.1

Is that intentional that you switched to a v0 major version tag there?

**Comment by iiinsomnia** [comment]
go 1.12.4 @andybons

**Comment by iiinsomnia** [comment]
@thepudds sorry, mistake in writing.
> go get A@v0.0.1

**Comment by jayconrod** [comment]
@iiinsomnia Could you post a small example repository with tags that demonstrates this problem? A sequence of commands would be helpful. And we still need the output of `go env`.

**Comment by iiinsomnia** [comment]
- go env
```sh
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/shenghui/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/workspace/go"
GOPROXY="https://goproxy.io"
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/j5/v0z7l6f53rs6lc9g2nyd37800000gn/T/go-build753395015=/tmp/go-build -gno-record-gcc-switches -fno-common"
```
- repository
```sh
github.com/iiinsomnia/yiigo@v3.1.2
```

**Comment by bcmills** [comment]
I think @thepudds has the write diagnosis.

@iiinsomnia, the mapping of commits to versions must not be changed once a version is published. That is a core property of Go modules: it is what makes builds reproducible and enables proxies and caching.

If there is a serious problem with `v1.0.1`, it's ok to remove the tag (so that new direct fetches will not be able to use that version), but the fix must be tagged `v1.0.2`, not `v1.0.1` again.

(Note that we do not currently have a mechanism to inform proxies that an existing version is no longer safe to use; that's #24031.)

**Comment by iiinsomnia** [comment]
thanks a lot !
